---
title: "Single Cell RNA Sequencing Analysis"
author: "Itay Nethanel Gal"
date: "1/5/23"
output: html_notebook
dataset: https://www.10xgenomics.com/resources/datasets/pbm-cs-from-a-healthy-donor-whole-transcriptome-analysis-3-1-standard-4-0-0
doc: https://docs.google.com/document/d/1fH_eAEC6MBjUyxtlzAF3WZK6KkYtdw4mOEiL6fK1AI4/edit?usp=sharing
GitHub repo: 
---

## Introduction

This notebook presents a walk through of Single Cell RNA Sequencing (scRNA-seq) data analysis. The advent of scRNA-seq technology has revolutionized the study of biology by enabling the examination of gene expression at the resolution of individual cells. By profiling gene expression across thousands of cells simultaneously, we can unravel the heterogeneity within a population of cells, identify novel cell types, and infer developmental trajectories.

We will use R, a language and environment for statistical computing and graphics, which is particularly well-suited to this type of data thanks to a rich ecosystem of libraries developed specifically for bioinformatics and computational biology. Some of the key packages we'll be utilizing in our analysis are `Seurat` for scRNA-seq data processing and analysis, `ggplot2` for data visualization, and `SingleCellExperiment` for managing single-cell experiment data.

The aim of this notebook is to provide a practical introduction to the analysis of scRNA-seq data, including:


Importing Libs:
```{r}
library(dplyr)
library(Seurat)
library(patchwork)
```

Getting the dataset:
```{r init}
# Load the PBMC dataset
getwd()
pbmc.data <- Read10X(data.dir = "Prod/")
# Initialize the Seurat object with the raw (non-normalized data).
pbmc <- CreateSeuratObject(counts = pbmc.data, project = "pbmc3k", min.cells = 3, min.features = 200)
pbmc
```
```{r}
# Lets examine a few genes in the first thirty cells
pbmc.data[c("CD3D","TCL1A","MS4A1"), 1:30]
```
- Here, we're going to extract the percentage of mitochondrial genes. Generally, dead cells present a high percentage of mitochondrial genes, so this analysis will help us      determine whether our cells are alive or dead
- We use the set of all genes starting with `MT-` as a set of mitochondrial genes

```{r mito, fig.height=7, fig.width=13}
# The [[ operator can add columns to object metadata. This is a great place to stash QC stats

pbmc[["percent.mt"]] <- PercentageFeatureSet(pbmc, pattern = "^MT-")
```

```{r qc, fig.height=7, fig.width=13}
# Show QC metrics for the first 5 cells
head(pbmc@meta.data, 5)
```
Here we are plotting the following information:

First Plot:
1. nFeature_RNA - Each black dot represent a cell and the Y axis represent the number of genes in each cell
2. nCount_RNA - Each black dot represent a cell and the Y axis represent the number of molecules in each cell
3. precent.mt - Each black dot represent a cell and the Y axis represent the percentage of mitochondrial genes (High means Bad - We will set cutoff to 5%-10%)

!!! Please see BeforeFiltering.png to see how the data looked like before the filtering !!!

```{r qc2, fig.height=7, fig.width=13}
#Visualize QC metrics as a violin plot
VlnPlot(pbmc, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
plot1 <- FeatureScatter(pbmc, feature1 = "nCount_RNA", feature2 = "percent.mt")
plot2 <- FeatureScatter(pbmc, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")
plot1
plot2
```

Filtering Out bad samples based on the Violin plots:
```{r}
pbmc <- subset(pbmc, subset = nFeature_RNA > 1000 & nFeature_RNA < 2500 & percent.mt < 10)
```


Adding Normalization layer to our data:
```{r}
pbmc <- NormalizeData(pbmc, normalization.method = "LogNormalize", scale.factor = 10000)
```
Showing the Violin Plots after Filtering and Normalization:
```{r qc2, fig.height=7, fig.width=13}
#Visualize QC metrics as a violin plot
VlnPlot(pbmc, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
plot1 <- FeatureScatter(pbmc, feature1 = "nCount_RNA", feature2 = "percent.mt")
plot2 <- FeatureScatter(pbmc, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")
plot1
plot2
```


Find VariableFeatures meaning find a genes that are highly expressed compare to other genes
The genes names we can see in the plot are the 10 most expressed genes among all the cells
```{r var_features, fig.height=5, fig.width=11}
pbmc <- FindVariableFeatures(pbmc, selection.method = 'vst', nfeatures = 2000)
# Identify the 10 most highly variable genes
top10 <- head(VariableFeatures(pbmc), 10)
# plot variable features with and without labels
plot1 <- VariableFeaturePlot(pbmc)
plot2 <- LabelPoints(plot = plot1, points = top10, repel = TRUE)
plot1 + plot2
```


Scaling the data to values between 0 to 1:
```{r regress, fig.height=7, fig.width=11, results='hide', eval = F}
all.genes <- rownames(pbmc)
pbmc <- ScaleData(pbmc, features = all.genes)
```
We are calculated PCs based on our genes and found PC1-PC5. 
More Information about it we can find here: https://www.youtube.com/watch?v=FgakZw6K1QQ (StatQuest video)
Notice that the number of PCs that was created is equal to the number of genes

```{r pca,results='hide'}
pbmc <- RunPCA(pbmc, features = VariableFeatures(object = pbmc))
```
Here each red dot represent a gene and the axis are PC1 and PC2. We can see that there is a cluster pattern that say that the expressiveness of the genes based on PC1 and PC2 are quite the same. Their SS(distance) to the origin (According to the video).
```{r pca_viz, message=TRUE}
# Examine and visualize PCA results a few different ways
print(pbmc[['pca']], dims = 1:5, nfeatures = 5)
VizDimLoadings(pbmc, dims = 1:2, reduction = 'pca')
DimPlot(pbmc, reduction = 'pca')
```

Plotting an HeatMap for PC1. We want the squares to be much distinguish and separate from each other which lead to more tendency to pick up this PC for our using
```{r single-heatmap}
DimHeatmap(pbmc, dims = 1, cells = 500, balanced = TRUE)
```

Here we are plotting the first 15 PCs. We can see that PC1-PC4 are the most significant therefore we will choose and continue with them

```{r multi-heatmap, fig.height=15, fig.width=9}
DimHeatmap(pbmc, dims = 1:15, cells = 500, balanced = TRUE)
```
JackStraw is another method in order to pick up PCs. This method display the P-Values for each PC and as we know low P-Value represent more significant result for the PC. We can see also here that PC1-PC4 represent the most significant P-Value results.
```{r}
# computation time
pbmc <- JackStraw(pbmc, num.replicate = 100)
pbmc <- ScoreJackStraw(pbmc, dims = 1:20)
JackStrawPlot(pbmc, dims = 1:15)
```

Elbow Plot also determine the significant of the PCs values based on std. We can see that The elbow is approximately at 4 therefore we will chose PC1-PC4 to continue
```{r elbow_plot, fig.height=6, fig.width=10}
ElbowPlot(pbmc)
```









